### PR TITLE
Pkg 1550 1.17.0 plus linux cuda

### PR DESCRIPTION
**Pull request by danpetry**
**Destination channel:** defaults

### Links

- [PKG-1550](https://anaconda.atlassian.net/browse/PKG-1550?atlOrigin=eyJpIjoiOGI0ZTNkMWYyNmZlNDc0ZThlOTBmODk4YmM1OTY3ZTkiLCJwIjoiaiJ9) 
- [Upstream repository](https://github.com/microsoft/onnxruntime/tree/v1.17.0)
- [Upstream changelog/diff](https://github.com/microsoft/onnxruntime/releases)

### Explanation of changes:

- Update to support CUDA 11.8 for the ticket; CPU versions updated alongside
- Tests enabled and various fixes related to that
- NB that support for CentOS 7 was dropped upstream (see changelog). Tests being run verifies that it's still ok.
- CUDA version is building successfully on a build machine, log [here](https://github.com/AnacondaRecipes/onnxruntime-feedstock/files/14638594/18buildafterfinaltidy.log.zip)
- see commit messages for details for changes.

[PKG-1550]: https://anaconda.atlassian.net/browse/PKG-1550?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by danpetry** [comment]
[CUDA build log file](https://github.com/AnacondaRecipes/onnxruntime-feedstock/files/14638594/18buildafterfinaltidy.log.zip)

File: recipe/meta.yaml [hunk 1]
```diff
@@ -39,26 +46,34 @@ requirements:
...
+    - cupti {{ cudatoolkit }}*        # [(ep_variant == "cuda") and (linux and x86_64)]
   run:
     - coloredlogs
     - packaging
     - protobuf
     - python
     - python-flatbuffers
     - sympy
-    - zstd
+    - {{ pin_compatible('zstd') }}
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml]: You could rely on the run exports here. Right now you are making the constraint more strict compared to the run_exports.

File: recipe/bld.bat [hunk 2]
```diff
@@ -5,17 +5,19 @@
...
     --build_wheel ^
     --config Release ^
     --update ^
     --build ^
+    --test ^
     --skip_submodule_sync
 if errorlevel 1 exit 1
 
-xcopy /S /Y /F build-ci\Release\Release\dist\onnxruntime-*.whl onnxruntime-%PKG_VERSION%-py3-none-any.whl*
+xcopy /S /Y /F build-ci\Release\dist\onnxruntime-*.whl onnxruntime-%PKG_VERSION%-py3-none-any.whl*
```
Comment (JeanChristopheMorinPerso) [inline recipe/bld.bat]: That copy is useless (and isn't done in conda-forge). It'll slow down the build.

File: recipe/build.sh [hunk 3]
```diff
@@ -48,7 +56,14 @@ ${PYTHON} tools/ci_build/build.py \
...
 
+if [[ "${ep_variant:-}" == "cuda" ]]; then
+    WHL_BASE_NAME="onnxruntime_gpu"
+else
+    WHL_BASE_NAME="onnxruntime"
+fi
 
-cp build-ci/Release/dist/onnxruntime-*.whl onnxruntime-${PKG_VERSION}-py3-none-any.whl
-${PYTHON} -m pip install onnxruntime-${PKG_VERSION}-py3-none-any.whl
+cp build-ci/Release/dist/${WHL_BASE_NAME}-*.whl ${WHL_BASE_NAME}-${PKG_VERSION}-py3-none-any.whl
```
Comment (JeanChristopheMorinPerso) [inline recipe/build.sh]: Same here, the copy isn't necessary.

**Comment by danpetry** [comment]
Thanks, due to the long build times I'll wait until I get another review before making changes.

Are these the only changes you want before an approval? if not, would be cool to get any other changes before I implement these, for the same reason, if you will be making more comments that is.

**Comment by danpetry** [comment]
CUDA build takes 24 hours

**Review by JeanChristopheMorinPerso — APPROVED** [review]
LGTM, on the condition that the comments are addressed before merging.

**Review by cbouss — APPROVED** [review]
LGTM, there is a lot of vendored stuff in the logs but that can be managed later.

**Comment by JeanChristopheMorinPerso** [comment]
>LGTM, there is a lot of vendored stuff in the logs but that can be managed later.

I saw that too and came to the same conclusion as you.

**Comment by danpetry** [comment]
Re protobuf, at least, I decided not to: https://github.com/conda-forge/onnxruntime-feedstock/issues/57#issuecomment-1518033552

File: recipe/meta.yaml (hunk 1 above)
Comment (danpetry) [reply recipe/meta.yaml]: done

File: recipe/bld.bat (hunk 2 above)
Comment (danpetry) [reply recipe/bld.bat]: done

File: recipe/build.sh (hunk 3 above)
Comment (danpetry) [reply recipe/build.sh]: done

**Pull request merged**